Improve prompt generation speed

calleerlandsson opened the pull request:
Generating and displaying the prompt in gitsh is currently painfully slow on my MacBook Air. This PR is an attempt to improve on this.

I started to investigate by printing all git commands passed to `Open3.capture3` in `GitRepository#git_output`:

```
"/usr/bin/env git rev-parse --git-dir"
"/usr/bin/env git rev-parse --git-dir"
"/usr/bin/env git symbolic-ref HEAD --short"
"/usr/bin/env git rev-parse --git-dir"
"/usr/bin/env git rev-parse --git-dir"
"/usr/bin/env git symbolic-ref HEAD --short"
"/usr/bin/env git rev-parse --git-dir"
"/usr/bin/env git rev-parse --git-dir"
"/usr/bin/env git symbolic-ref HEAD --short"
"/usr/bin/env git rev-parse --git-dir"
"/usr/bin/env git rev-parse --git-dir"
"/usr/bin/env git status --porcelain"
"/usr/bin/env git status --porcelain"
"/usr/bin/env git config --get-color gitsh.color.modified yellow"
"/usr/bin/env git rev-parse --git-dir"
"/usr/bin/env git rev-parse --git-dir"
"/usr/bin/env git status --porcelain"
"/usr/bin/env git status --porcelain"
```

Generating the prompt on my system took ~1.1 seconds.

I observed that most of the calls to external git commands were duplicated.

I then added `puts` statements in many of the methods in `Prompter` and `PromptColor` in order to find out where the calls were made. I went on to reduce this number by introducing memoization in a way that, to my knowledge, doesn't affect the behavior of the program.

The three commits in this PR are the results of this. It brings the generation time down to 0.3 seconds and the following calls to external git commands:

```
"/usr/bin/env git rev-parse --git-dir"
"/usr/bin/env git status --porcelain"
"/usr/bin/env git config --get-color gitsh.color.modified yellow"
"/usr/bin/env git symbolic-ref HEAD --short"
```

In doing this I exposed a new `status` method on `GitRepository`. I didn't figure out a good way to test this new method. It is currently implicitly tested through the `has_untracked_files?` and `has_modified_files?` specs. I considered the following options:
1. Duplicate the existing tests but in the exercise, call `status.has_untracked_files?` instead of `has_untracked_files?`. I don't think the duplicated tests would provide much value.
2. Create a new spec for `GitRepository::Status` passing in example `git status --porcelain` output and replace the existing specs with specs that test that the original methods delegate to the status object. Since the existing specs tests the git repository on an integration level by altering a real git repository and the new specs would be hard to implement this way, this option seems to reduce the value of the test suite.

I'd love some input on this.

In my last commit I also inline the `PromptColor` method within the `Prompter` to make use of the cache inside it. Only having one method and no private helpers, it seemed like the `PromptColor` didn't really pull it's own weight so I think this is a good trade off. I'm also unsure of the value provided by the `PromptColor` spec since it mostly duplicated the conditional logic of the implementation. Therefor I didn't add any similar tests for `Prompter`.

georgebrock commented on lib/gitsh/git_repository.rb:
```diff
@@ -8,7 +8,8 @@ def initialize(env)
     def initialized?
-      git_dir && File.exist?(git_dir)
+      dir = git_dir
+      dir && File.exist?(dir)
```
What do you think of this as an alternative?

```ruby
def initialized?
  File.exist?(git_dir)
rescue TypeError
  false
end
```

mike-burns replied:
I'm confused in all these suggestions -- not sure what I'm missing -- but what is the `TypeError` that would be raised? Can `git_output` produce `nil`?

georgebrock replied:
The `git_dir && …` in the original implementation appeared to be there to guard against `nil`, so I assumed that it could.

However, it seems I'm wrong. It can be the empty string, but since `File.exist?('')` returns `false` we can probably simplify this even further.

georgebrock commented on lib/gitsh/prompter.rb:
```diff
@@ -11,25 +10,33 @@ class Prompter
-        '%w' => clear_color,
-        '%#' => terminator
-      })
+      reset_chached_repo_info
```
Typo: s/chached/cached/

georgebrock replied:
We generally avoid creating state outside of `#initialize` methods in this codebase, and often create private classes that can hold the state they need to process a specific call (e.g. in the `Gitsh::Completer#call` method). Does it make sense to follow that pattern here too?

Something like this maybe?

```ruby
def prompt
    Prompt.new(
      env: env,
      repo_initialized: …,
      branch_name: …,
      repo_status: …,
    ).to_s
end
```

georgebrock commented:
Thanks so much for this! The speed improvement is great.

> In doing this I exposed a new `status` method on `GitRepository`. I didn't figure out a good way to test this new method. It is currently implicitly tested through the `has_untracked_files?` and `has_modified_files?` specs. I considered the following options:
>
> 1. Duplicate the existing tests but in the exercise, call `status.has_untracked_files?` instead of `has_untracked_files?`. I don't think the duplicated tests would provide much value.
> 2. Create a new spec for `GitRepository::Status` passing in example `git status --porcelain` output and replace the existing specs with specs that test that the original methods delegate to the status object. Since the existing specs tests the git repository on an integration level by altering a real git repository and the new specs would be hard to implement this way, this option seems to reduce the value of the test suite.
>
> I'd love some input on this.

I prefer the second option. As long as there's a true integration spec somewhere that exercises a couple of the prompt variants, I think it's OK to test this at the unit level with stubbed collaborators. [`spec/integration/prompt_spec.rb`](https://github.com/thoughtbot/gitsh/blob/master/spec/integration/prompt_spec.rb) seems to cover the basics already. Maybe we just need another example in there to demonstrate the sigil changing when we modify files?

> In my last commit I also inline the `PromptColor` method within the `Prompter` to make use of the cache inside it. Only having one method and no private helpers, it seemed like the `PromptColor` didn't really pull it's own weight so I think this is a good trade off. I'm also unsure of the value provided by the `PromptColor` spec since it mostly duplicated the conditional logic of the implementation. Therefor I didn't add any similar tests for `Prompter`.

I think there was some value in these specs. Particularly, the order of the `elsif` cases matters, and without that test I don't think we exercise all of the possibilities anywhere anymore.

How would you feel about changing the `Gitsh::PromptColor` class to accept a `Status` object as an argument to its `#status_color` method? That way, it could still be tested separately, but it could also use the cached state.

calleerlandsson commented:
I'm sorry for dropping the ball on this one.

I have now rebased on master and updated the PR according to your comments:

- I added specs for `Gitsh::GitRepository::Status` and a sigil example to `spec/integration/prompt_spec.rb`
- I brought back `PromptColor` and its specs and made `status_color` accept a `Status` object as its parameter.
- I extracted a `Prompt` class encapsulating the cached state required for prompt generation.
- Simplified the implementation of `#initialized?`.

I also:

- Moved `GitRepository#initialized?` into `GitRepository::Status` to limit the number of parameters to pass to `PromptColor#status_color`.
- Realized that the `GitRepository` methods now delegating to the `Status` object and the `Environment` methods delegating to these were used only by objects with access to a `Status` object. Because of this, I removed the delegating methods.

georgebrock commented:
This is great, thanks Calle!
